### ISSUE TITLE
Hermitian Transpose Syntax

### ISSUE THREAD

**Issue by sclel016** [numpy/numpy#13797]
Hi,

For people doing signal processing, a concise way to express the Hermitian Transpose would lead to more readable code. Currently, the syntax for a Hermitian transpose of an array is

`A.conj().T`

The syntax used in the matrix class `.H` should be ported to numpy.array. A similar comment could be made for the matrix inverse and `.I`.

ref: [StackExchange](https://stackoverflow.com/questions/26932461/conjugate-transpose-operator-h-in-numpy)

**Comment by mhvk** [comment]
Like the idea for `.H` in principle, but fear we're a little stuck in `.T` already not quite doing what one would want: it transposes *all* axes rather than just swapping the last two, which means that for stacks of matrices, it is generally not what you want (see #9530).  A suggestion there was `.MT` (for matrix-transpose`).

I'm less sure about the inverse - mostly as a property suggests a cheap operation that cannot really fail, which does not hold for the inverse.

**Comment by sclel016** [comment]
Fair comment about `.I`, I hadn't thought of the failure mode. I was under the impression that for a multidimensional transpose, it's proper notation to specify which two dimensions are being transposed. Maybe a syntax similar to `A.T(1,3)`  might be helpful with a default `A.T` swapping the last two?

**Comment by mhvk** [comment]
@sclel016 - the problem is that `A.T` is already defined and always swaps the order of *all* axes - I'd love for it to be just the last two, but that would break other people's code, so we cannot do that. Hence, the idea for a new property. But if we have a new name for that, we probably should name the hermitian transpose in a similar way.

**Comment by sclel016** [comment]
In order to avoid breaking existing code, what are your thoughts on `A.MT` as suggested for transpose and `A.HT` for Hermitian transpose.

**Comment by mhvk** [comment]
That seems quite neat.

Normally, new features are proposed on the mailing list. Do you want to send a message to ask people what they think? See https://mail.python.org/mailman/listinfo/numpy-discussion

**Comment by sclel016** [comment]
I can do that. Thanks!

**Comment by StanczakDominik** [comment]
That indeed sounds like a pretty neat idea :)

**Comment by eric-wieser** [comment]
[Mailing list post](http://numpy-discussion.10968.n7.nabble.com/Syntax-Improvement-for-Array-Transpose-td47200.html)

**Comment by rgommers** [comment]
For reference, `.H` was discussed and rejected several times in the past. gh-8882 has rationale. I propose to close this issue.

**Comment by seberg** [comment]
Sounds fair, if there is something remotely close to a consensus on the mailing list (and I doubt it is possible for `.H`) we can reopen. My guess is the only thing that could in principle fly is `.mT`, but having both `.mT` and `.T` seems strange as well. If someone proposes a serious `.H` I am all ears, but first implement that conjugate dtype (and do not recoil in horror) ;).

**Comment by mhvk** [comment]
I'm sorry, but reopening until the mailing list discussion actually ends. I feel quite strongly this is a feature worth having, and do not recall any consensus.

**Comment by mhvk** [comment]
p.s. The fact that is has been asked for so repeatedly should perhaps be seen for what it is: something people really wish numpy had.

**Comment by rgommers** [comment]
I agree, no need to close just yet. 

> p.s. The fact that is has been asked for so repeatedly should perhaps be seen for what it is: something people really wish numpy had.

It does feel more like we're re-hashing old discussions without addressing or even acknowledging previous ones.....

**Comment by mhvk** [comment]
> It does feel more like we're re-hashing old discussions without addressing or even acknowledging previous ones.....

A consequence partially of some discussions not leading to a conclusion and those conclusions that are reached not being written down. But on the upside of course also of the people involved changing and thus bringing in different priorities and ideas.

**Comment by mhvk** [comment]
p.p.s. I'll stop now, but I think we perhaps fail to acknowledge how much the `Matrix` class was in fact used - we're deprecating it saying all it brings is better done with regular arrays but do not provide even minimal infrastructure. E.g., how is it possible that numpy does not have a `matrix_transpose` function? (I write this wondering even why I didn't put a PR in myself...) I guess cecause of `Matrix` and lack of thinking beyond a single matrix.

**Comment by eric-wieser** [comment]
> how is it possible that numpy does not have a `matrix_transpose` function

I hid an undocumented one at `np.linalg.transpose` that uses the same broadcasting rules as the other linalg functions. You're welcome ;)

**Comment by eric-wieser** [comment]
@mhvk: Elaborated on that thought in #13835

**Comment by seberg** [comment]
I am not quite convinced that we cannot change `.T` (at least in the sense of deprecation) myself, but it seemed like enough pushback that it needs  bigger write up. I am fairly convinced that `.H` being a view is a fairly complex thing.

One thing I have to agree with Marten: We had said the same about other things (most noisily/annoying the integer deprecation for example), and while it was very annoying, I think nowadays the consensus is that it was a good idea. Yet in the code there were clear comments saying "unfortunately there is no way we can pull that off", I am sure it was discussed a lot before (heck `__index__` was invented for numpy after all).

A new stacks of matrices class might be a good mid point. I think I might go with more errors. Even the old matrix had an `.A` attribute or so to get an array view fast. So if you want funny indexing that is not super clear.... just let the user go through that. I would probably prefer it to be prototyped outside of numpy first, but we can put it into the numpy repository.

**Comment by charris** [comment]
I don't recall being in opposition, and I also think the current transpose is not what we want. Now there is the problem of consistency, transpose can be a view and we probably don't want the Hermitian conjugate to be a view, but having a shorthand way to do it would be useful. I think it could even be done as a generalized ufunc.

**Comment by mhvk** [comment]
I thought of the "complex conjugate" dtype a little more: I think it is not difficult to implement in itself, and it might help speed up/save memory on (some) complex-valued calculations of very large arrays (but should be probably be treated as the hack it is, i.e., invisible from the user to the extent possible). But one thing that would be agreed on beforehand is what to do with `cc.imag`. I don't see any solution but for that to return a copy (probably best read-only, since writing to it would not change the original array), or to error.

**Comment by seberg** [comment]
I do think the complex conjugate dtype is not very difficult as such. I am more scared that hiding it well may end up ugle and other smaller traps crop up. And yes `cc.imag` would somewhat require another negative real to do?

**Comment by mhvk** [comment]
> And yes `cc.imag` would somewhat require another negative real to do?

That would indeed solve the "no-copy-for-a-property" fear. But that would seem to increase the likelihood of ugly traps further down for very little benefit. It might then be better to just error on `.imag`.

Anyway, my sense is that it does not help in terms of the perceived problem with `.H` doing a copy rather than returning a view, so while I still think it is a really neat idea, we might as well leave it out of the discussion here.

**Comment by cjblocker** [comment]
To summarize the mailing list discussion, both [the more recent one](http://numpy-discussion.10968.n7.nabble.com/Syntax-Improvement-for-Array-Transpose-td47200.html) and [the one from several years ago](http://numpy-discussion.10968.n7.nabble.com/add-H-attribute-td34474.html), there is a general consensus for a `.H` that can be implemented without a copy. 

As was originally described [here on the mailing list](http://numpy-discussion.10968.n7.nabble.com/add-H-attribute-tp34474p34655.html),  and discussed above, this could be implemented with a conjugate dtype. In addition to saving the copy, this would also open the possibility of using BLAS routines that implement A.H@B, offering further performance improvements. All ufuncs could be supported by a safe cast to a complex dtype.

If no one else was intending to work on this, I'd be happy to contribute my time to implement this. In regards to what to do with `cc.imag`, I agree that an error would be best, and easy for people to work around. It could then be addressed later if there is a large demand for some other functionality.

**Comment by sclel016** [comment]
That's really great to hear Cameron, I'm glad you're willing to take this on! Can I make a suggestion of adding optional arguments to specify which two dimensions should be hermitian transposed? Also, what would be the expected behavior of `.H` when using an array of real numbers?

**Comment by cjblocker** [comment]
Adding arguments is hard, since `A.H(0,1)` would be parsed as `C = A.H; C(0,1)`, so if `.H` were a property, then `C` would be an array and `A.H(0,1)` would be calling `C`. `.H` could be a method instead of a property, but there was resistance on the mailing list from `.H` acting differently than `.T` outside of conjugation. That is also one of the reasons `.H` should return a view instead of a copy. 

There was a discussion on the mailing list about a `.mH`/`.mT` that would just transpose the last two dimensions. I'm not sure what the verdict on that was, but the trend of the conversation towards the end suggested that level of specialization may be delegated to array subclasses like the np.matrix replacement Eric linked to above.

For arrays with a real dtype, I would expect `.H` to functionally just be an alias for `.T`.

**Comment by adeak** [comment]
@eric-wieser is `linalg.transpose` undocumented because it's trivial to implement, because you fear it can cause confusion _vis-à-vis_ `numpy.transpose`, or some other reason? It would seem useful still, especially if referenced in an issue ;)

**Comment by jakirkham** [comment]
This seems like a duplicate of issue ( https://github.com/numpy/numpy/issues/5763 ). Though this one is more up-to-date. Maybe one of these should be closed?

**Comment by seberg** [comment]
Yes, `.mT` is unblocked since all previous resistance to it is gone as far as I am aware.  Maybe even changing `.T` is unblocked at least in the sense of deprecating confusing forms (although I can imagine it to be a noisy one).

`.H` does have the copy problems.  Yes, you can write a complex conjugate derivative of the normal complex dtype.  But, unless you also implement at least basic arithmetics for that dtype, I am not convinced it is super useful?  And having an odd representation for complex has its own annoyances, for example you won't be able to just share easily with Cython through the buffer protocol.
(Which you may decide is OK, if a `arr.H` array can have weird properties since they will go away almost immediately in most cases.)

OTOH, I am not sure that `.H` doing the copy is *that* bad, it could even be a read-only to not give an idea of being able to modify the original array (and in theory make a view more reasonable in the future).

**Comment by rgommers** [comment]
> Yes, you can write a complex conjugate derivative of the normal complex dtype

If you want to look at an approach like that, PyTorch implements a "complex conjugate bit" that makes `.H` zero-copy. It's not exactly trivial implementation-wise, but it does save performance/memory and as far as I know folks were happy with it.

**Comment by charris** [comment]
Setting a bit on a view could be useful, but all the complex functions would need to be aware of it.
